cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -0,0 +1,706 @@
+                        uint32_t n_clusters,
+                        uint32_t dim,
+                        const float* dataset,
+                        size_t n_rows,
+                        uint32_t* labels,
+                        raft::distance::DistanceType metric,
+                        rmm::cuda_stream_view stream,
+                        rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<float> distances(n_rows * n_clusters, stream, mr);
+
+  float alpha;
+  float beta;
+  switch (metric) {
+    case raft::distance::DistanceType::InnerProduct: {
+      alpha = -1.0;
+      beta  = 0.0;
+    } break;
+    case raft::distance::DistanceType::L2Expanded:
+    case raft::distance::DistanceType::L2Unexpanded: {
+      rmm::device_uvector<float> sqsum_centers(n_clusters, stream, mr);
+      rmm::device_uvector<float> sqsum_data(n_rows, stream, mr);
+      utils::dots_along_rows(n_clusters, dim, centers, sqsum_centers.data(), stream);
+      utils::dots_along_rows(n_rows, dim, dataset, sqsum_data.data(), stream);
+      utils::outer_add(
+        sqsum_data.data(), n_rows, sqsum_centers.data(), n_clusters, distances.data(), stream);
+      alpha = -2.0;
+      beta  = 1.0;
+    } break;
+    // NB: update the description of `knn::ivf_flat::build` when adding here a new metric.
+    default: RAFT_FAIL("The chosen distance metric is not supported (%d)", int(metric));
+  }
+  linalg::gemm(handle,
+               true,
+               false,
+               n_clusters,
+               n_rows,
+               dim,
+               &alpha,
+               centers,
+               dim,
+               dataset,
+               dim,
+               &beta,
+               distances.data(),
+               n_clusters,
+               stream);
+  utils::argmin_along_rows(n_rows, n_clusters, distances.data(), labels, stream);
+}
+
+/**
+ * @brief Suggest a minibatch size for kmeans prediction.
+ *
+ * This function is used as a heuristic to split the work over a large dataset
+ * to reduce the size of temporary memory allocations.
+ *
+ * @param n_clusters number of clusters in kmeans clustering
+ * @param n_rows dataset size
+ * @return a suggested minibatch size
+ */
+constexpr auto calc_minibatch_size(uint32_t n_clusters, size_t n_rows) -> uint32_t
+{
+  n_clusters              = std::max<uint32_t>(1, n_clusters);
+  uint32_t minibatch_size = (1 << 20);
+  if (minibatch_size > (1 << 28) / n_clusters) {
+    minibatch_size = (1 << 28) / n_clusters;
+    minibatch_size += 32;
+    minibatch_size -= minibatch_size % 64;
+  }
+  minibatch_size = uint32_t(std::min<size_t>(minibatch_size, n_rows));
+  return minibatch_size;
+}
+
+/**
+ * @brief Given the data and labels, calculate cluster centers and sizes in one sweep.
+ *
+ * Let S_i = {x_k | x_k \in dataset & labels[k] == i} be the vectors in the dataset with label i.
+ *   On exit centers_i = normalize(\sum_{x \in S_i} x), where `normalize` depends on the distance
+ * type.
```

Comment:
Shall be updated since normalization is moved out, and it would help if the weighted average is explained:

```suggestion
 * Let S_i = {x_k | x_k \in dataset & labels[k] == i} be the vectors in the dataset with label i.
 *   On exit centers_i = (\sum_{x \in S_i} x + w_i * center_i) / (|S_i| + w_i), where  w_i = reset_counters ?  0 : cluster_size[i]. In other words, the updated cluster centers are a weighted average of the existing cluster center, and the coordinates of the points labeled with i.
```